libraries/Microsoft.Bot.Builder.Dialogs/Debugging/SourceMap.cs

```diff
@@ -1,24 +1,91 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.Bot.Builder.Dialogs.Debugging
 {
     /// <summary>
     /// A simple ISourceMap of objects -> SourceRange.
     /// </summary>
-    public class SourceMap : ConcurrentDictionary<object, SourceRange>, ISourceMap
+    public class SourceMap : ISourceMap
     {
         /// <summary>
         /// Initializes a read-only new instance of the <see cref="SourceMap"/>.
         /// </summary>
         public static readonly SourceMap Instance = new SourceMap();
+        
+        private readonly object _gate = new object();
+        private readonly Dictionary<object, SourceRange> _items = new Dictionary<object, SourceRange>(ReferenceEquality<object>.Instance);
+        private readonly Dictionary<SourceRange, object> _reverseLookup = new Dictionary<SourceRange, object>(new SourceRangeEqualityComparer());
 
-        /// <inheritdoc/>
-        public void Add(object item, SourceRange range)
+        void ISourceMap.Add(object item, SourceRange range)
         {
-            this.AddOrUpdate(item, (key) => range, (key, prev) => range);
+            if (range.Path != null && !Path.IsPathRooted(range.Path))
+            {
+                throw new ArgumentOutOfRangeException(range.Path);
+            }
+
+            lock (_gate)
+            {
+                if (_reverseLookup.TryGetValue(range, out object foundRef))
+                {
+                    _items.Remove(foundRef);
+                    _reverseLookup.Remove(range);
+                }
+
+                _items.Add(item, range);
+                _reverseLookup.Add(range, item);
+            }
+        }
+
+        bool ISourceMap.TryGetValue(object item, out SourceRange range)
+        {
+            if (item != null)
+            {
+                lock (_gate)
+                {
+                    return _items.TryGetValue(item, out range);
+                }
+            }
+
+            range = default;
+            return false;
+        }
+    }
+
+#pragma warning disable SA1402 // File may only contain a single type
+    internal sealed class ReferenceEquality<T> : IEqualityComparer<T>
+#pragma warning restore SA1402 // File may only contain a single type
+    {
+        public static readonly IEqualityComparer<T> Instance = new ReferenceEquality<T>();
+
+        private ReferenceEquality()
+        {
+        }
+
+        bool IEqualityComparer<T>.Equals(T x, T y) => ReferenceEquals(x, y);
+
+        int IEqualityComparer<T>.GetHashCode(T obj) => RuntimeHelpers.GetHashCode(obj);
+    }
+
+#pragma warning disable SA1402 // File may only contain a single type
+    internal class SourceRangeEqualityComparer : EqualityComparer<SourceRange>
+#pragma warning restore SA1402 // File may only contain a single type
+    {
+        public override bool Equals(SourceRange x, SourceRange y)
+        {
+            return x.Equals(y);
```

Comment:
Given that SourceRange already overrides Equals and hashCode, we might not need this comparer, in which case we can delete after testing to confirm.